### PR TITLE
Optimize `MemoryMarshal.Cast` for same type

### PR DESCRIPTION
**Pull request by xtqqczze**
Follow-up to https://github.com/dotnet/runtime/pull/99835.

### Diffs

- https://github.com/MihuBot/runtime-utils/issues/348
- https://github.com/MihuBot/runtime-utils/issues/349

Diffs looks mostly positive although there are some large code size regressions due to enablement of more inlining.

cc: @stephentoub

### CODE REVIEW

**Comment by EgorBo** [comment]
Didn't we discuss that this fast path makes no sense? in https://github.com/dotnet/runtime/pull/99835#discussion_r1526776477

**Comment by jakobbotsch** [comment]
Just echoing what @EgorBo said, due to that stall mentioned there, it's not clearly a win even if the codegen "looks" better. For example:
```csharp
[Benchmark]
public void BenchBase()
{
    Span<int> s;
    for (int i = 0; i < 10000; i++)
    {
        TestBase(_values.AsSpan(0, i), out s);
    }
}

[Benchmark]
public void BenchDiff()
{
    Span<int> s;
    for (int i = 0; i < 10000; i++)
    {
        TestDiff(_values.AsSpan(0, i), out s);
    }
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static void TestBase<T>(Span<T> s, out Span<int> s2) where T : struct
{
    s2 = CastBase<T, int>(s);
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static void TestDiff<T>(Span<T> s, out Span<int> s2) where T : struct
{
    s2 = CastDiff<T, int>(s);
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static unsafe Span<TTo> CastDiff<TFrom, TTo>(Span<TFrom> span)
    where TFrom : struct
    where TTo : struct
{
    if (typeof(TFrom) == typeof(TTo))
        return *(Span<TTo>*)&span;

    // Use unsigned integers - unsigned division by constant (especially by power of 2)
    // and checked casts are faster and smaller.
    uint fromSize = (uint)Unsafe.SizeOf<TFrom>();
    uint toSize = (uint)Unsafe.SizeOf<TTo>();
    uint fromLength = (uint)span.Length;
    int toLength;
    if (fromSize == toSize)
    {
        // Special case for same size types - `(ulong)fromLength * (ulong)fromSize / (ulong)toSize`
        // should be optimized to just `length` but the JIT doesn't do that today.
        toLength = (int)fromLength;
    }
    else if (fromSize == 1)
    {
        // Special case for byte sized TFrom - `(ulong)fromLength * (ulong)fromSize / (ulong)toSize`
        // becomes `(ulong)fromLength / (ulong)toSize` but the JIT can't narrow it down to `int`
        // and can't eliminate the checked cast. This also avoids a 32 bit specific issue,
        // the JIT can't eliminate long multiply by 1.
        toLength = (int)(fromLength / toSize);
    }
    else
    {
        // Ensure that casts are done in such a way that the JIT is able to "see"
        // the uint->ulong casts and the multiply together so that on 32 bit targets
        // 32x32to64 multiplication is used.
        ulong toLengthUInt64 = (ulong)fromLength * (ulong)fromSize / (ulong)toSize;
        toLength = checked((int)toLengthUInt64);
    }

    return MemoryMarshal.CreateSpan(ref Unsafe.As<TFrom, TTo>(ref MemoryMarshal.GetReference(span)), toLength);
}

[MethodImpl(MethodImplOptions.AggressiveInlining)]
public static unsafe Span<TTo> CastBase<TFrom, TTo>(Span<TFrom> span)
    where TFrom : struct
    where TTo : struct
{
    // Use unsigned integers - unsigned division by constant (especially by power of 2)
    // and checked casts are faster and smaller.
    uint fromSize = (uint)Unsafe.SizeOf<TFrom>();
    uint toSize = (uint)Unsafe.SizeOf<TTo>();
    uint fromLength = (uint)span.Length;
    int toLength;
    if (fromSize == toSize)
    {
        // Special case for same size types - `(ulong)fromLength * (ulong)fromSize / (ulong)toSize`
        // should be optimized to just `length` but the JIT doesn't do that today.
        toLength = (int)fromLength;
    }
    else if (fromSize == 1)
    {
        // Special case for byte sized TFrom - `(ulong)fromLength * (ulong)fromSize / (ulong)toSize`
        // becomes `(ulong)fromLength / (ulong)toSize` but the JIT can't narrow it down to `int`
        // and can't eliminate the checked cast. This also avoids a 32 bit specific issue,
        // the JIT can't eliminate long multiply by 1.
        toLength = (int)(fromLength / toSize);
    }
    else
    {
        // Ensure that casts are done in such a way that the JIT is able to "see"
        // the uint->ulong casts and the multiply together so that on 32 bit targets
        // 32x32to64 multiplication is used.
        ulong toLengthUInt64 = (ulong)fromLength * (ulong)fromSize / (ulong)toSize;
        toLength = checked((int)toLengthUInt64);
    }

    return MemoryMarshal.CreateSpan(ref Unsafe.As<TFrom, TTo>(ref MemoryMarshal.GetReference(span)), toLength);
}
```

|    Method |     Mean |    Error |   StdDev |
|---------- |---------:|---------:|---------:|
| BenchBase | 12.96 us | 0.109 us | 0.102 us |
| BenchDiff | 45.36 us | 0.146 us | 0.122 us |

**Comment by xtqqczze** [comment]
@jakobbotsch Thanks for the BenchmarkDotNet example, is this issue specific to XArch?

```log
BenchmarkDotNet v0.13.12, macOS Sonoma 14.4 (23E214) [Darwin 23.4.0]
Apple M2, 1 CPU, 8 logical and 8 physical cores
.NET SDK 8.0.203
  [Host]     : .NET 8.0.3 (8.0.324.11423), Arm64 RyuJIT AdvSIMD
  Job-UYGVGG : .NET 8.0.3 (8.0.324.11423), Arm64 RyuJIT AdvSIMD

Runtime=.NET 8.0  Toolchain=net80  
```

| Method    | Mean     | Error     | StdDev    |
|---------- |---------:|----------:|----------:|
| BenchBase | 9.612 us | 0.0354 us | 0.0314 us |
| BenchDiff | 9.627 us | 0.0495 us | 0.0463 us |

File: src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs [hunk 1]
```diff
@@ -122,6 +122,9 @@ public static unsafe ReadOnlySpan<byte> AsBytes<T>(ReadOnlySpan<T> span)
             if (RuntimeHelpers.IsReferenceOrContainsReferences<TTo>())
                 ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(TTo));
 
+            if (typeof(TFrom) == typeof(TTo))
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs:125]: This change would be a regression for shared generic code. `TFrom` and `TTo` may require generic dictionary lookups.

The identical types should be covered by `if (fromSize == toSize)` path below. Is there anything preventing the JIT from generating the optimal code for that path?

**Pull request closed without merging**